### ISSUE TITLE
ctest fails on appveyor

### ISSUE THREAD

**Issue by fedorov** [QIICR/dcmqi#26]
There is some ctest issue that prevents tests from running on appveyor. Same command seems to work fine on circleci and travis ...

![image](https://cloud.githubusercontent.com/assets/313942/16467343/ecd9699a-3e14-11e6-823c-2ea128634a3c.png)

**Comment by fedorov** [comment]
And on circleci too!

![image](https://cloud.githubusercontent.com/assets/313942/16522732/8c71d876-3f6d-11e6-9694-c7d2281a33bf.png)

**Comment by fedorov** [comment]
but not on travis!

![image](https://cloud.githubusercontent.com/assets/313942/16522755/a3e25a1c-3f6d-11e6-81e8-771ebf8df9dc.png)

**Comment by fedorov** [comment]
Setting the path to include lib folders of ITK/DCMTK/zlib didn't resolve the problem

https://github.com/QIICR/dcmqi/blob/master/appveyor.yml#L36

**Comment by fedorov** [comment]
![image](https://cloud.githubusercontent.com/assets/313942/16745886/f8d7ae44-4785-11e6-96be-f45a981c38d8.png)

**Comment by fedorov** [comment]
@pieper can you point me to the location where library paths are set for Slicer launcher? I am not sure where the libraries end up on windows. There may be some other problem too, but I think paths are still not correct (I don't know where compiled libs are for SlicerExecutionModel and dcmqi; ITK and dcmtk are downloaded from a packaged release).

**Comment by fedorov** [comment]
Current test command:

```
cd c:\dcmqi\dcmqi-build\dcmqi-build && set PATH=%PATH%;c:\DCMTK-install\lib;c:\ITK-install\lib;c:\zlib-install\lib && pwd && dir && tree && ctest -VV -C Release -D ExperimentalTest --no-compress-output
```

failing test

![image](https://cloud.githubusercontent.com/assets/313942/17833875/27a9ad6c-66f8-11e6-96fe-611e35c2c9ed.png)


failing dashboard: https://ci.appveyor.com/project/fedorov/dcmqi/build/0.0.234

**Comment by pieper** [comment]
@fedorov are you looking for this?
https://github.com/Slicer/Slicer/blob/master/SuperBuild/External_CTKAPPLAUNCHER.cmake

or did you mean you want to use the applauncher for dcmqi utilities?  If so
you can look here:

http://commontk.org/index.php/Tools:_Application_launcher

But I thought the goal was to build static so why use the launcher at all?

**Comment by fedorov** [comment]
> But I thought the goal was to build static so why use the launcher at all?

All I need is to figure out why ctest is failing on appveyor.

**Comment by pieper** [comment]
Do you have a windows build of your own that works?

**Comment by fedorov** [comment]
No, I never compiled it on anything other than my mac.

I will add a command to run the executable directly to see the error, outside ctest, and if that doesn't work, I will look into setting up a windows build.

@che85 I believe you have a windows dev setup - maybe you could take a look what's going on there?

**Comment by pieper** [comment]
If needed I can try a windows build.

**Comment by che85** [comment]
There is that tower (where Windows is installed) standing here in my office, but I am not using it at all. 

@pieper If you already have a windows machine running, that might be easier I think.

**Comment by fedorov** [comment]
@che85 I was thinking about your windows tablet, but sounds like you are not using it for development

**Comment by che85** [comment]
I changed my mind about developing on windows (which is sometimes really painful). I am using my tablet mostly for Photoshop, but rarely for any development activities.

**Comment by pieper** [comment]
I did a build now on windows 7, visual studio 2008, cmake 3.4.

The everything builds fine, but the executables don't run because they
can't find the dlls.

I tried setting DCMQI to build static, but the ITK and DCMTK dlls are still
not found.

If I copy all the dcmqi *.exe files and all the ITK and DCMTK *.dll files
into a new folder then the dcmqi executables run.

So to make the tests run on AppVeyor on windows the options would be:

1) set the PATH variable to include the ITK and DCMTK build directories
either in CMake or by creating a wrapper .bat file.  This is probably the
easiest option.

2) copy all the files into a new folder and execute from there.

3) configure and use a ctkAppLauncher

4) figure out how to build DCMTK and ITK static and pass in the flags
during the SuperBuild.  This may not completely work since ITK uses dynamic
libraries for the IOFactories and they will need to be discoverable.

5) wait for Jc and all to make the installer and then always run the
install step before running the tests.  This might be the cleanest because
it would test an actual dcmqi installation directory rather than a build
tree and that's one step closer to a real user environment.

**Comment by fedorov** [comment]
Steve, thank you for investigating! I think it would be good to have PATH set as a temporary option. Are you sure it is ITK and DCMTK dlls not found, and not SlicerExecutionModel and dcmqi? I do set the paths to those manually, and since I do it for a package, I thought I know what those paths should be.

Can you also check and let me know what are the paths where SlicerExecutionModel and dcmqi dlls are saved?

**Comment by fedorov** [comment]
If you could make a PR with the fix, this of course would be ideal!

**Comment by pieper** [comment]
I'm pretty sure it's just the ITK and DCMTK paths since the executables ran
once I moved them to the right directory.

I guess I could do a PR for the quick fix.  Not sure I'll get to it today
though.

BTW, all the jsonlint and ajv tests fail too.  Are they intended to be
installed via the superbuild or is it assumed they will be in the path?  It
might make sense to include instructions where to get them in an error
message when the test fails.

**Comment by fedorov** [comment]
> I'm pretty sure it's just the ITK and DCMTK paths since the executables ran
once I moved them to the right directory.

I guess this means I somehow messed up this line: https://github.com/QIICR/dcmqi/blob/master/appveyor.yml#L35

> I guess I could do a PR for the quick fix.  Not sure I'll get to it today
though.

No rush, any time is good. I just think it is most efficient to do this with a windows build handy, otherwise a lot of inefficient commits to poke appveyor.

> BTW, all the jsonlint and ajv tests fail too.  

Right, I saw that, but also don't know why. These tools are installed in the pre-build step: https://github.com/QIICR/dcmqi/blob/master/appveyor.yml#L12-L13. I will add to the instructions.

**Comment by pieper** [comment]
Okay - I see now how AppVeyor is set up - well maybe you only need to
change the paths to be c:\DCMTK-install\bin and c:\ITK\install\bin because
that is where the dlls are located (next to the .exe files for runtime -
the lib folder is for libraries used to build new exes.)

Getting it to test right on a windows development machine would be similar
but maybe can be on the list for Jc and all.

**Comment by fedorov** [comment]
@pieper nice, thanks - binary tests are running now! 

now the only examples failing are json tests. If you know where npm installs binaries, please let me know. I added `c:\Users\appveyor\AppData\Roaming\npm` to the path, but it didn't help.

**Comment by pieper** [comment]
nice!


For ajv you could dry:

%HOME%\AppData\Roaming\npm\ajv.cmd

HTH!

**Comment by fedorov** [comment]
I do have `%HOME%\AppData\Roaming\npm` in the PATH, and on the windows machine I tried, it works (ajv and jsonlint binaries are in that folder on my windows). I don't want to specify ajv.cmd directly in the test, since it will not work on other platforms.

**Comment by pieper** [comment]
It probably depends on the shell from which you execute ajv - a native
windows environment may require ajv.cmd but you could do that with a test
and setting a variable in cmake if it came to that.

**Comment by fedorov** [comment]
@pieper both `ajv` and `ajv.cmd` appear to work in appveyor `cmd` shell. Looks like we need help from the cmake geniuses!

@jcfr @msmolens do you have any idea about what is going on here:

* we are testing on appveyor using npm-installed packages
* packages are installed in this line: https://github.com/QIICR/dcmqi/blob/master/appveyor.yml#L12-L13
* these lines https://github.com/QIICR/dcmqi/blob/master/appveyor.yml#L35-L36 both work fine on appveyor: https://ci.appveyor.com/project/fedorov/dcmqi/build/0.0.251#L1203 (i.e., npm-installed packages can be used fine)
* the path to the `ajv` package used by `ctest` is identical to the one that works when invoked directly from the `cmd` shell; here is the location where `ctest` is invoked: https://github.com/QIICR/dcmqi/blob/master/appveyor.yml#L37, and this is the where it is being run by appveyor: https://ci.appveyor.com/project/fedorov/dcmqi/build/0.0.251#L1662

Bottom line: `ajv` works when invoked directly from the shell, but not when called by `ctest`.

Why is `ctest` failing to run the test with "Not Run/BAD_COMMAND" status?

![image](https://cloud.githubusercontent.com/assets/313942/17896111/64dd49d8-691d-11e6-98e1-8e498bf9a2ff.png)

**Comment by fedorov** [comment]
@pieper I don't understand why, but indeed the cmd version works from ctest, as demonstrated in https://github.com/QIICR/dcmqi/pull/79. I didn't even try before, because non-cmd version works from the command line, and I thought something else is going on, but apparently I was wrong.

**Comment by pieper** [comment]
I think you should just have a check for windows at the top of the cmake
file and then set a variable called AJV that is either ajv or ajv.cmd
depending on the platform.  Then every time you use it you use ${AJV}
instead.  You could do the same for ${JSONLINT}.

**Comment by fedorov** [comment]
Yes, definitely - I'll wait a bit for @IlyaFinkelshteyn, as the original contributor of #79, to fix this, and otherwise will fix myself.
